src/LaunchDarkly.ServerSdk/Internal/Model/Clause.cs

```diff
@@ -63,19 +63,32 @@ private static PreprocessedData Preprocess(Operator op, IEnumerable<LdValue> val
                         return new PreprocessedValue();
                     }
                 }).ToImmutableList()
-            };
+            );
 
-        internal struct PreprocessedData
+        internal readonly struct PreprocessedData
         {
-            internal ImmutableHashSet<LdValue> ValuesAsSet { get; set; }
-            internal ImmutableList<PreprocessedValue> Values { get; set; }
+            internal ImmutableHashSet<LdValue> ValuesAsSet { get; }
+            internal ImmutableList<PreprocessedValue> Values { get; }
+
+            public PreprocessedData(ImmutableHashSet<LdValue> valuesAsSet = null, ImmutableList<PreprocessedValue> values = null)
+            {
+                ValuesAsSet = valuesAsSet;
+                Values = values;
+            }
         }
 
-        internal struct PreprocessedValue
+        internal readonly struct PreprocessedValue
         {
-            internal Regex Regex { get; set; }
-            internal DateTime? DateTime { get; set; }
-            internal SemanticVersion? SemVer { get; set; }
+            internal Regex Regex { get; }
+            internal DateTime? DateTime { get; }
+            internal SemanticVersion? SemVer { get; }
+
+            public PreprocessedValue(Regex regex = null, DateTime? dateTime = null, SemanticVersion? semVer = null)
```

Comment:
Only ever setting one at a time so I just matched the equivalent getter/setter interface that existed